contracts/hts-precompile/AtomicHTS.sol

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.20;
+
+import "./IHRC.sol";
+import "./HederaTokenService.sol";
+import "./IHederaTokenService.sol";
+
+/**
+ * @dev This contract contains multiple examples highlighting the utilization of the 
+ * HIP-551: Batch transactions using HTS calls via HederaTokenService Precompile contract.
+ * The batches rolls multiple HTS calls in one transaction that passes ACID test 
+ * (atomicity, consistency, isolation, and durability).
+ * Read more about HIP-551 at https://hips.hedera.com/hip/hip-551
+ */
+contract AtomicHTS is HederaTokenService {
+    /// events ///
+    event BatchAssociateGrantKYCTransfer(int associateResponseCode, int grantKYCResponseCode, int transferTokenResponseCode);
+    event BatchApproveAssociateGrantKYCTransferFrom(int transferTokenResponseCode, int approveResponseCode, int associateResponseCode, int grantKYCResponseCode, int transferFromResponseCode);
+    event BatchUnfreezeGrantKYCTransferFreeze(int unfreezeTokenResponseCode, int grantKYCResponseCode, int transferTokenResponseCode, int freezeTokenResponseCode);
+    event BatchWipeMintTransfer(int wipeTokenResponseCode, int mintTokenResponseCode, int transferTokenResponseCode);
+    event BatchMintUnfreezeGrantKYCTransferFreeze(int mintTokenResponseCode, int unfreezeTokenResponseCode, int grantKYCResponseCode, int freezeTokenResponseCode);
+    event BatchAssociateMintGrantTransfer(int associateResponseCode, int mintTokenResponseCode, int grantKYCResponseCode, int transferTokenResponseCode);
+
+
+    /**
+     * @dev associates, grant token KYC, and send an amount of fungible token to a receiver.
+     * - associateToken() -> grantTokenKYC() -> transferToken()
+     */
+    function batchAssociateGrantKYCTransfer(address token, address sender, address receiver, int64 amount) external {
+        (int associateResponseCode) = HederaTokenService.associateToken(receiver, token);
+        require(associateResponseCode == HederaResponseCodes.SUCCESS, "Failed to associate token.");
+
+        /// @notice an account needs to be granted the KYC of the HTS token for it to receive the token
+        (int grantKYCResponseCode) = HederaTokenService.grantTokenKyc(token, receiver);
```

Comment:
Similar concept here. If there's a response code for already granted we should allow that and not fail. Same thing further down

Reply:
Pushed in the fix. Also I figured that this concept is only applicable for associate methods. Other actions like grantKYC or freeze can be overlapped.